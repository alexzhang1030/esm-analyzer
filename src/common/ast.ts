import type { ASTNode, ASTNodeLocation, t, WalkCallback } from '@/types'
import { walk } from 'estree-walker'

export function walkAST(node: ASTNode, handlers: {
  enter?: WalkCallback
  leave?: WalkCallback
}): ASTNode {
  return walk(node, handlers)
}

function withOffsetContent(offsetContent: string, d: { line: number, column: number, index: number }) {
  if (!offsetContent.length)
    return d
  const lines = offsetContent.split('\n')
  const numNewlines = lines.length - 1
  if (numNewlines > 0)
    d.line += numNewlines
  else
    d.column += offsetContent.length
  return {
    line: d.line,
    column: d.column,
    index: d.index + offsetContent.length,
  }
}

// offset accepts offsetContent and offset number
export function getASTNodeLocation(node: ASTNode, offsetContent: string = ''): ASTNodeLocation {
  const { start: _s, end: _e } = node.loc!
  return {
    start: withOffsetContent(offsetContent, { ..._s, index: node.start! }),
    end: withOffsetContent(offsetContent, { ..._e, index: node.end! }),
  }
}

// ast is
export function isImportDeclaration(node: ASTNode): node is t.ImportDeclaration {
  return node.type === 'ImportDeclaration'
}

export function isExportNamedDeclaration(node: ASTNode): node is t.ExportNamedDeclaration {
  return node.type === 'ExportNamedDeclaration'
}

export function isExportDefaultDeclaration(node: ASTNode): node is t.ExportDefaultDeclaration {
  return node.type === 'ExportDefaultDeclaration'
}

export function isExportAllDeclaration(node: ASTNode): node is t.ExportAllDeclaration {
  return node.type === 'ExportAllDeclaration'
}

export function isVariableDeclaration(node: ASTNode): node is t.VariableDeclaration {
  return node.type === 'VariableDeclaration'
}

export function isIdentifier(node: ASTNode): node is t.Identifier {
  return node.type === 'Identifier'
}
