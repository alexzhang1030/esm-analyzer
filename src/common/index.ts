export * from './ast'
export * from './lang'
export * from './parse'
export * from './variable'
