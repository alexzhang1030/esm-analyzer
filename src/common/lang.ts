import type { AcceptableLang } from '..'

const REGEX_LANG_TS = /.?[mc]?tsx?$/
const REGEX_LANG_JSX = /.?[jt]sx$/
const REGEX_ACCEPTABLE_LANG = /.?[mc]?[jt]sx?$/

export const isTS = (lang: string) => REGEX_LANG_TS.test(lang)
export const isJSX = (lang: string) => REGEX_LANG_JSX.test(lang)
export const isAcceptableLang = (filename: string) => REGEX_ACCEPTABLE_LANG.test(filename)

export function getLangByFileName(filename: string) {
  return filename.split('.').pop() as AcceptableLang
}
