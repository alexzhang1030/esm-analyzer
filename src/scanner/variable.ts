import type { ResolveVariableDeclaration, VariableType } from '@/common'
import {
  getASTNodeLocation, isIdentifier,
  isVariableDeclaration, resolveVariableDeclarationValue,
} from '@/common'
import type { ASTNode, ASTNodeLocation, t } from '@/types'
import { loop } from '@/utils'

export interface ScanVariableDeclarationResult {
  loc: ASTNodeLocation
  kind: t.VariableDeclaration['kind']
  name: string
  init: ResolveVariableDeclaration
}

export interface ScanVariableDeclarationConfig {
  includeType?: VariableType[]
  excludeType?: VariableType[]
}

export function scanVariableDeclaration(node: ASTNode, config?: ScanVariableDeclarationConfig): ScanVariableDeclarationResult[] | null {
  if (!isVariableDeclaration(node))
    return null
  const result: ScanVariableDeclarationResult[] = []
  loop(node.declarations, (declaration) => {
    // TODO: handle more cases, currently only handle `identifier`
    // TODO: handle delay init, e.g. `var` and `let`
    if (!isIdentifier(declaration.id))
      return
    const value = {
      loc: getASTNodeLocation(declaration),
      kind: node.kind,
      name: declaration.id.name,
      init: resolveVariableDeclarationValue(declaration.init, config),
    }
    if (!value.init)
      return
    result.push(value)
  })
  return result
}
