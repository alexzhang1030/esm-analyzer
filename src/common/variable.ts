import type { Primitive } from 'type-fest'
import { isIdentifier } from './ast'
import type { ASTNode } from '@/types'

type PrimitiveValue =
  | Primitive
  | null | undefined

export interface PrimitiveVariableValue {
  type: 'StringLiteral' | 'NumericLiteral' | 'BooleanLiteral' | 'NullLiteral' | 'ObjectExpression' | 'ArrayExpression'
  value: PrimitiveValue | ResolveVariableDeclaration
}

export interface ObjectExpressionVariableValue {
  type: 'ObjectExpression'
  value: {
    [key: string]: ResolveVariableDeclaration
  }
}

export interface ArrayExpressionVariableValue {
  type: 'ArrayExpression'
  value: ResolveVariableDeclaration[]
}

export interface CallExpressionVariableValue {
  type: 'CallExpression'
  callee: string
  arguments: ResolveVariableDeclaration[]
}

export type ResolveVariableDeclaration =
  | PrimitiveVariableValue
  | CallExpressionVariableValue
  | ObjectExpressionVariableValue
  | ArrayExpressionVariableValue
  | null

export function resolveVariableDeclarationValue(node?: ASTNode | null): ResolveVariableDeclaration {
  if (!node)
    return null
  switch (node.type) {
    case 'StringLiteral':
    case 'NumericLiteral':
    case 'BooleanLiteral':
      return {
        type: node.type,
        value: node.value,
      }
    case 'NullLiteral':
      return {
        type: node.type,
        value: null,
      }
    case 'ObjectExpression':
      return {
        type: node.type,
        value: node.properties.reduce((result, property) => {
          if (!property || property.type !== 'ObjectProperty')
            return result
          if (!isIdentifier(property.key))
            return result
          if (!property.value)
            return result
          result[property.key.name] = resolveVariableDeclarationValue(property.value)
          return result
        }, {} as ObjectExpressionVariableValue['value']),
      }
    case 'ArrayExpression':
      return {
        type: node.type,
        value: node.elements.map((item) => {
          return resolveVariableDeclarationValue(item)
        }),
      }
    case 'CallExpression':
      if (!node.callee || !isIdentifier(node.callee))
        return null
      return {
        type: node.type,
        callee: node.callee.name,
        arguments: node.arguments.map((item) => {
          return resolveVariableDeclarationValue(item)
        }),
      }
  }
  return null
}
