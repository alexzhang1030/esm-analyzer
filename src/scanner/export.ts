import type { ResolveVariableDeclaration } from '@/common'
import { getASTNodeLocation, isExportAllDeclaration, isExportDefaultDeclaration, isExportNamedDeclaration, isIdentifier, resolveVariableDeclarationValue } from '@/common'
import type { ASTNode, ASTNodeLocation, t } from '@/types'
import { loop } from '@/utils'

export type ScanExportType =
  | 'ExportNamedDeclaration'
  | 'ExportAllDeclaration'
  | 'ExportDefaultDeclaration'

export interface ScanExportNamedResult {
  type: 'ExportNamedDeclaration'
  kind: t.VariableDeclaration['kind']
  declarations: {
    name: string
    init: ResolveVariableDeclaration
  }[]
}

export interface ScanExportAllResult {
  type: 'ExportAllDeclaration'
  source: string
}

export interface ScanExportDefaultIdentifierResult {
  type: 'ExportDefaultDeclaration'
  subType: 'Identifier'
  id: string
}

export interface ScanExportDefaultObjectResult {
  type: 'ExportDefaultDeclaration'
  subType: 'ObjectExpression'
  properties: {
    key: string
    value: ResolveVariableDeclaration
  }[]
}

export type ScanExportResult = (
  | ScanExportNamedResult
  | ScanExportAllResult
  | ScanExportDefaultIdentifierResult
  | ScanExportDefaultObjectResult
) & {
  loc: ASTNodeLocation
}

export interface ScanExportConfig {
  includeType?: ScanExportType[]
  excludeType?: ScanExportType[]
}

export function scanExport(node: ASTNode, config?: ScanExportConfig): ScanExportResult | null {
  if (
    !isExportNamedDeclaration(node)
    && !isExportDefaultDeclaration(node)
    && !isExportAllDeclaration(node)
  )
    return null
  if (config?.includeType && !config.includeType.includes(node.type))
    return null
  if (config?.excludeType && config.excludeType.includes(node.type))
    return null
  if (node.type === 'ExportNamedDeclaration') {
    if (!node.declaration || node.declaration.type !== 'VariableDeclaration')
      return null
    const vars: ScanExportNamedResult['declarations'] = []
    loop(node.declaration.declarations, (declaration) => {
      if (!isIdentifier(declaration.id))
        return
      const result = {
        name: declaration.id.name,
        init: resolveVariableDeclarationValue(declaration.init),
      }
      if (!result.init)
        return
      vars.push(result)
    })
    return {
      type: 'ExportNamedDeclaration',
      declarations: vars,
      kind: node.declaration.kind,
      loc: getASTNodeLocation(node),
    }
  }
  else if (node.type === 'ExportAllDeclaration') {
    return {
      type: 'ExportAllDeclaration',
      source: node.source.value,
      loc: getASTNodeLocation(node),
    }
  }
  else {
    if (node.declaration.type === 'Identifier') {
      return {
        type: 'ExportDefaultDeclaration',
        subType: 'Identifier',
        id: node.declaration.name,
        loc: getASTNodeLocation(node),
      }
    }
    else if (node.declaration.type === 'ObjectExpression') {
      return {
        type: 'ExportDefaultDeclaration',
        subType: 'ObjectExpression',
        loc: getASTNodeLocation(node),
        properties: node.declaration.properties.map((property) => {
          if (property.type !== 'ObjectProperty' || !isIdentifier(property.key))
            return null
          return {
            key: property.key.name,
            value: resolveVariableDeclarationValue(property.value),
          }
        }).filter(Boolean) as ScanExportDefaultObjectResult['properties'],
      }
    }
  }
  return null
}
