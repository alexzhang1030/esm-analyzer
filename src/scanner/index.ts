import type { ScanImportConfig } from './import'
import { scanImport } from './import'
import type { AcceptableLang } from '@/types'
import { loadScanner } from '@/common'

interface ScanConfig {
  import?: ScanImportConfig
}

const defaultConfig: ScanConfig = {
  import: undefined,
}

export function scan(code: string, lang: AcceptableLang, config: ScanConfig = defaultConfig) {
  const [imports] = loadScanner(code, lang, [node => scanImport(node, config.import)])
  return {
    imports,
  }
}

export * from './import'
export * from './variable'
