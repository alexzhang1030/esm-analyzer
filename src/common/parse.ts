import { parse, type ParserOptions, type ParserPlugin } from '@babel/parser'
import { isJSX, isTS } from './lang'

export function getBabelParsePlugins(lang: string) {
  const plugins: ParserPlugin[] = []
  if (isTS(lang))
    plugins.push('typescript')
  if (isJSX(lang))
    plugins.push('jsx')
  return plugins
}

export function babelParse(code: string, lang: string) {
  const options: ParserOptions = {
    plugins: getBabelParsePlugins(lang),
    sourceType: 'module',
  }
  return parse(code, options)
}
