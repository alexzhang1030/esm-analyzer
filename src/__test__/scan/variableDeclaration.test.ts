import { loadScanner } from '@/common'
import { scanVariableDeclaration } from '@/scanner/variable'

describe('variableDeclaration', () => {
  test('primitive', () => {
    const code = `
      const a = 1
      const b = '2'
      let c = true
      var d = null
    `
    const result = loadScanner(code, 'js', [node => scanVariableDeclaration(node)])[0]
    expect(result).toMatchSnapshot()
  })
  test('function call', () => {
    const code = `
      const foo = ref(1)
      const bar = computed(() => 1)
      const baz = reactive({ a: 1 })
      const baz2 = ref([1, 2, 3])
    `
    const result = loadScanner(code, 'js', [node => scanVariableDeclaration(node)])[0]
    expect(result).toMatchSnapshot()
  })
})
