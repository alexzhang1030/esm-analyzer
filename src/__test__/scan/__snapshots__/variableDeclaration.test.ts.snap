// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`variableDeclaration > function call 1`] = `
[
  {
    "init": {
      "arguments": [
        {
          "type": "NumericLiteral",
          "value": 1,
        },
      ],
      "callee": "ref",
      "type": "CallExpression",
    },
    "kind": "const",
    "loc": {
      "end": {
        "column": 24,
        "index": 25,
        "line": 2,
      },
      "start": {
        "column": 12,
        "index": 13,
        "line": 2,
      },
    },
    "name": "foo",
  },
  {
    "init": {
      "arguments": [
        null,
      ],
      "callee": "computed",
      "type": "CallExpression",
    },
    "kind": "const",
    "loc": {
      "end": {
        "column": 35,
        "index": 61,
        "line": 3,
      },
      "start": {
        "column": 12,
        "index": 38,
        "line": 3,
      },
    },
    "name": "bar",
  },
  {
    "init": {
      "arguments": [
        {
          "type": "ObjectExpression",
          "value": {
            "a": {
              "type": "NumericLiteral",
              "value": 1,
            },
          },
        },
      ],
      "callee": "reactive",
      "type": "CallExpression",
    },
    "kind": "const",
    "loc": {
      "end": {
        "column": 36,
        "index": 98,
        "line": 4,
      },
      "start": {
        "column": 12,
        "index": 74,
        "line": 4,
      },
    },
    "name": "baz",
  },
  {
    "init": {
      "arguments": [
        {
          "type": "ArrayExpression",
          "value": [
            {
              "type": "NumericLiteral",
              "value": 1,
            },
            {
              "type": "NumericLiteral",
              "value": 2,
            },
            {
              "type": "NumericLiteral",
              "value": 3,
            },
          ],
        },
      ],
      "callee": "ref",
      "type": "CallExpression",
    },
    "kind": "const",
    "loc": {
      "end": {
        "column": 33,
        "index": 132,
        "line": 5,
      },
      "start": {
        "column": 12,
        "index": 111,
        "line": 5,
      },
    },
    "name": "baz2",
  },
]
`;

exports[`variableDeclaration > primitive 1`] = `
[
  {
    "init": {
      "type": "NumericLiteral",
      "value": 1,
    },
    "kind": "const",
    "loc": {
      "end": {
        "column": 17,
        "index": 18,
        "line": 2,
      },
      "start": {
        "column": 12,
        "index": 13,
        "line": 2,
      },
    },
    "name": "a",
  },
  {
    "init": {
      "type": "StringLiteral",
      "value": "2",
    },
    "kind": "const",
    "loc": {
      "end": {
        "column": 19,
        "index": 38,
        "line": 3,
      },
      "start": {
        "column": 12,
        "index": 31,
        "line": 3,
      },
    },
    "name": "b",
  },
  {
    "init": {
      "type": "BooleanLiteral",
      "value": true,
    },
    "kind": "let",
    "loc": {
      "end": {
        "column": 18,
        "index": 57,
        "line": 4,
      },
      "start": {
        "column": 10,
        "index": 49,
        "line": 4,
      },
    },
    "name": "c",
  },
  {
    "init": {
      "type": "NullLiteral",
      "value": null,
    },
    "kind": "var",
    "loc": {
      "end": {
        "column": 18,
        "index": 76,
        "line": 5,
      },
      "start": {
        "column": 10,
        "index": 68,
        "line": 5,
      },
    },
    "name": "d",
  },
]
`;
